### ISSUE TITLE
StrictRedisCluster fails to run cluster commands when slots are not initialised

### ISSUE THREAD

**Issue by kunal-bajpai** [Grokzen/redis-py-cluster#246]
While using StrictRedisCluster, it seems like an undocumented pre-requisite is having a cluster with all hashslots assigned.

Without that, cluster commands fail to run
```
>>> s=StrictRedisCluster(startup_nodes=[{'host':'127.0.0.1','port':2951}],decode_responses=True)
>>> s.cluster_info()
{}
>>> s.cluster_slots()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/kajaybaj/.pex/install/redis_py_cluster-1.3.4-py2.py3-none-any.whl.dbcde0484cbf5a0e89b6bf98ba40b8fed1fdce10/redis_py_cluster-1.3.4-py2.py3-none-any.whl/rediscluster/client.py", line 595, in cluster_slots
    return self.execute_command('CLUSTER SLOTS')
  File "/home/kajaybaj/.pex/install/redis_py_cluster-1.3.4-py2.py3-none-any.whl.dbcde0484cbf5a0e89b6bf98ba40b8fed1fdce10/redis_py_cluster-1.3.4-py2.py3-none-any.whl/rediscluster/utils.py", line 101, in inner
    return func(*args, **kwargs)
  File "/home/kajaybaj/.pex/install/redis_py_cluster-1.3.4-py2.py3-none-any.whl.dbcde0484cbf5a0e89b6bf98ba40b8fed1fdce10/redis_py_cluster-1.3.4-py2.py3-none-any.whl/rediscluster/client.py", line 317, in execute_command
    node = self.determine_node(*args, **kwargs)
  File "/home/kajaybaj/.pex/install/redis_py_cluster-1.3.4-py2.py3-none-any.whl.dbcde0484cbf5a0e89b6bf98ba40b8fed1fdce10/redis_py_cluster-1.3.4-py2.py3-none-any.whl/rediscluster/client.py", line 297, in determine_node
    return [self.connection_pool.nodes.random_node()]
  File "/home/kajaybaj/.pex/install/redis_py_cluster-1.3.4-py2.py3-none-any.whl.dbcde0484cbf5a0e89b6bf98ba40b8fed1fdce10/redis_py_cluster-1.3.4-py2.py3-none-any.whl/rediscluster/nodemanager.py", line 115, in random_node
    key = random.choice(list(self.nodes.keys()))
  File "/export/apps/python/3.6/lib/python3.6/random.py", line 257, in choice
    raise IndexError('Cannot choose from an empty sequence') from None
IndexError: Cannot choose from an empty sequence
```
It seems reasonable to want to use StrictRedisCluster to initialise a cluster (addslots, cluster replicate, etc) but it doesn't work because of the following loop that iterates over results of CLUSTER SLOTS 
https://github.com/Grokzen/redis-py-cluster/blob/2cbe4a5ba859571e1acb86f33f202904a4d18cb1/rediscluster/nodemanager.py#L193

With some overrides of the routing mechanism, I've come up with a class that works with any cluster. It would be nice to have this support upstream.
```
class RedisCluster(StrictRedisCluster):

    def __init__(self, *args, **kwargs) -> None:  # type: ignore
        self.NODES_FLAGS['CLUSTER INFO'] = 'all'

    def determine_node(self, *args, **kwargs):  # type: ignore
        node = super().determine_node(*args, **kwargs)

        if node is None:
            node_flag = self.nodes_flags.get(args[0])
            if node_flag == 'all':
                node = [self.connection_pool.nodes.make_node_obj(host=h['host'] or '127.0.0.1',
                                                                 port=h['port'],
                                                                 server_type='slave' if 'slave' in h['flags'] else 'master')[0] for h in self.cluster_nodes()
```

Basically, this logic relies on `CLUSTER NODES` command to populate node list rather than `CLUSTER SLOTS`

**Comment by Grokzen** [comment]
@kunal-bajpai I am not really sure how i would impelment this cleanly becuase creating node objects that is not derived from the cluster itself is a very big step away from the current cluster discovery mechanisms. I do get the usecase for using the lib for creating and managing a cluster. But the current commands is more directed towards managing a operating cluster and not to be a part of setting up one.

Switching from cluster nodes to cluster slots will basically not get you anywhere becuase they kinda contain the same information. If you do not have any slots or a cluster created both will be empty and you have not gained anything to help your case. Basically what you need is to NOT have any kind of cluster discovery or to have your startup_nodes to be the populated nodes with a empty slots cache when your cluster is not initialized properly.

I will put it on the backlog but i have to think more to weed this out and how to implement it cleanly.

**Comment by kunal-bajpai** [comment]
My suggestion to use CLUSTER NODES was only to fetch a list of all connected nodes. This is so that commands like CLUSTER INFO, that have nothing to do with hashslots, have a target node list initialised irrespective of slot assignment.

Currently the node list is compiled while iterating through the results for CLUSTER SLOTS (NodeManager.initialize). I suggest you separate the node list initialisation from node-to-slot mapping so that cluster setup commands continue to work even without assigned hashslots.

**Comment by kunal-bajpai** [comment]
In my opinion, ideally, startup_nodes should only be used to figure out the target for CLUSTER NODES. Thereafter, the node list from CLUSTER NODES should be used to populate nodes.

**Comment by Grokzen** [comment]
This issue is stale and i basically have no other implementation mechanism that i want to implement. This lib is currently not really made for building and setting up a cluster as that should either be done manually by building and running your own `Redis` class instances and point them to your nodes in your cluster and then run management commands to them manually. I will not reshape the current discovery mechanism and basically say that you need to have a cluster up and running for this lib to be able to connect into it and consume it. I will say that possibly the new future ClusterManagement feature maybe can facilitate this but it is a bit of time out before it will drop into the master code.
